### PR TITLE
PLT-2998 Check that outgoing webhooks and slash commands have unique trigger words

### PR DESCRIPTION
**Pull request by DavidLu1997**
Previously, outgoing webhooks and slash commands did not have checks to ensure that triggers were unique.

For slash commands, triggers cannot be duplicated.
For outgoing webhooks, two webhooks in the same channel cannot share any triggers and callback URLs. (Approved by LB)
 - For example, a webhook with the triggers `{"cats", "dogs"}` and `{"http://pets.com"}` cannot exist at the same time as  `{"cats", "fish"}` and `{"http://pets.com"}`. But `{"cats", "dogs"}` and `{"http://pets.com"}` can exist at the same time as `{"derp"}` and `{"http://pets.com"}`

### CODE REVIEW

File: api/command.go [hunk 1]
```diff
@@ -86,6 +86,32 @@ func listCommands(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(model.CommandListToJson(commands)))
 }
 
+func getCommands(c *Context) []*model.Command {
+	commands := make([]*model.Command, 0, 32)
+	seen := make(map[string]bool)
```
Comment (coreyhulen) [inline api/command.go]: Why is seen only for built in commands?  would make sense for it to be the other way around.

**Comment by DavidLu1997** [comment]
The function returns both in the end, the code was copy pasted from the listCommands function, but instead of writing a HttpResponse, it returns an array. The unit tests check for two commands with the trigger `trigger`.

File: api/command.go [hunk 2]
```diff
@@ -86,6 +86,32 @@ func listCommands(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(model.CommandListToJson(commands)))
 }
 
+func getCommands(c *Context) []*model.Command {
```
Comment (crspeller) [inline api/command.go]: getAllCommands? Is this what this function is doing?

File: api/command.go [hunk 3]
```diff
@@ -86,6 +86,32 @@ func listCommands(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	commands := make([]*model.Command, 0, 32)
+	seen := make(map[string]bool)
+	for _, value := range commandProviders {
+		cpy := *value.GetCommand(c)
```
Comment (crspeller) [inline api/command.go]: `value`? `cpy`? I can't tell what's going on.

File: api/command.go [hunk 4]
```diff
@@ -86,6 +86,32 @@ func listCommands(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		if cpy.AutoComplete && !seen[cpy.Id] {
+			cpy.Sanitize()
+			seen[cpy.Trigger] = true
+			commands = append(commands, &cpy)
+		}
+	}
+
+	if *utils.Cfg.ServiceSettings.EnableCommands {
```
Comment (crspeller) [inline api/command.go]: Why are we just checking this now?

File: model/utils.go [hunk 5]
```diff
@@ -420,3 +420,20 @@ func IsSafeLink(link *string) bool {
 
 	return true
 }
+
+func Intersection(arr1, arr2 StringArray) StringArray {
```
Comment (crspeller) [inline model/utils.go]: Intersection of what?

**Comment by jwilander** [comment]
Will work with DL to resubmit

**Pull request closed without merging**